baseplate/observers/prometheus.py

```diff
@@ -115,7 +119,80 @@ def on_child_span_created(self, span: Span) -> None:
 
 class PrometheusClientSpanObserver(SpanObserver):
     def __init__(self) -> None:
-        logger.debug("PrometheusClientSpanObserver not implemented")
+        self.tags: Dict[str, Any] = {}
+        self.start_time: Optional[int] = None
+        self.metrics: Optional[
+            Union[PrometheusHTTPClientMetrics]
+        ] = None  # Add PrometheusThriftClientMetrics when implemented
+
+    def on_set_tag(self, key: str, value: Any) -> None:
+        self.tags[key] = value
+
+    def get_prefix(self) -> str:
+        return f"{self.protocol}_client"
+
+    def get_labels(self) -> Dict[str, str]:
+        return self.tags
+
+    @property
+    def protocol(self) -> str:
+        return self.tags.get("protocol", "unknown")
+
+    def set_metrics_by_protocol(self) -> None:
+        if self.protocol == "http":
+            self.metrics = PrometheusHTTPClientMetrics()
+        elif self.protocol == "thrift":
+            # self.metrics = PrometheusThriftClientMetrics()
+            logger.warning("PrometheusThriftClientMetrics not implemented.")
+        else:
+            logger.warning(
+                "No valid protocol set for Prometheus metric collection, metrics won't be collected. Expected 'http' or 'thrift' protocol. Actual protocol: %s",
+                self.protocol,
+            )
+        return
+
+    def on_start(self) -> None:
+        self.set_metrics_by_protocol()
+        if self.metrics is None:
+            logger.warning(
+                "No metrics set for Prometheus metric collection. Metrics will not be exported correctly."
+            )
+            return
+        self.start_time = time.perf_counter_ns()
+        self.metrics.active_requests_metric(self.tags).inc()
+
+    def on_incr_tag(self, key: str, delta: float) -> None:
+        pass
+
+    def on_finish(self, exc_info: Optional[_ExcInfo]) -> None:
+        if self.metrics is None:
+            logger.warning(
+                "No metrics set for Prometheus metric collection. Metrics will not be exported correctly."
+            )
+            return
+
+        self.tags["success"] = "true"
+        if exc_info is not None:
+            self.tags["exception_type"] = exc_info[1].__class__.__name__
+            self.tags["success"] = "false"
+
+        self.metrics.active_requests_metric(self.tags).dec()
+        self.metrics.requests_total_metric(self.tags).inc()
+        if self.start_time is not None:
+            elapsed_ns = time.perf_counter_ns() - self.start_time
+            self.metrics.latency_seconds_metric(self.tags).observe(
+                elapsed_ns / NANOSECONDS_PER_SECOND
+            )
+
+    def on_child_span_created(self, span: Span) -> None:
+        observer: Optional[SpanObserver] = None
+        if isinstance(span, LocalSpan):
+            observer = PrometheusLocalSpanObserver()
```

Comment:
Why is it important to differentiate local spans here?

Reply:
I haven't been able to get somebody to explain to me what LocalSpans are used for, and which metrics we should expose for them if any. Cullen was going to setup a meeting about that I think.
So this comes from the original implementation I was working from, to answer your question :-)

Reply:
Gotcha – there's a bit of info [here](https://baseplate.readthedocs.io/en/stable/api/baseplate/index.html#baseplate.ServerSpan.make_child) that I found useful, not sure if you've seen that already.

Not sure if we need/want to instrument them by default, but should we at least implement the `on_child_span_created` function so HTTP calls nested in local spans are instrumented?

e.g. I _think_ something like this would not record metrics for the HTTP call since `PrometheusLocalSpanObserver` doesn't attach 

```python
with baseplate.server_context("test") as context:
   with context.span.make_child('my super expensive operation', local=True) as span:
      sleep(5)
      span.context.myclient.get('http://example.org')
```

Reply:
Thanks for bringing that up. It would technically have been fixed when merging the local span implementation... but this is probably better to implement here.

Reply:
Thanks for adding this 👍